File: HisabCore/Sources/HisabCore/Reconciliation.swift
import Foundation

/// Projection of a stored transaction for matching purposes.
public struct ReconTxn: Sendable {
    public let id: UUID
    public let date: Date
    public let amountPaise: Int64
    public let direction: Direction
    public let reference: String?

    public init(id: UUID, date: Date, amountPaise: Int64, direction: Direction, reference: String?) {
        self.id = id
        self.date = date
        self.amountPaise = amountPaise
        self.direction = direction
        self.reference = reference
    }
}

public enum MatchTier: String, Codable, Sendable {
    case reference, amountDate
}

public struct MatchPair: Sendable, Equatable {
    public let appID: UUID
    public let bankID: UUID
    public let tier: MatchTier

    public init(appID: UUID, bankID: UUID, tier: MatchTier) {
        self.appID = appID
        self.bankID = bankID
        self.tier = tier
    }
}

public struct ReconciliationResult: Sendable {
    public let matches: [MatchPair]
    public let appUnmatched: [UUID]
    public let bankOnly: [UUID]
}

public enum Reconciler {
    /// Tier 1: exact reference match. Tier 2: same (amount, direction) within ±window days,
    /// closest date first; each bank transaction consumed at most once.
    public static func reconcile(app: [ReconTxn], bank: [ReconTxn], dateWindowDays: Int = 2) -> ReconciliationResult {
        var matches: [MatchPair] = []
        var consumedBank = Set<UUID>()
        var unmatchedApp: [ReconTxn] = []

        // Tier 1 — reference
        var bankByRef: [String: ReconTxn] = [:]
        for txn in bank {
            if let ref = txn.reference, !ref.isEmpty, bankByRef[ref] == nil {
                bankByRef[ref] = txn
            }
        }
        for txn in app {
            if let ref = txn.reference, !ref.isEmpty, let hit = bankByRef[ref], !consumedBank.contains(hit.id) {
                matches.append(MatchPair(appID: txn.id, bankID: hit.id, tier: .reference))
                consumedBank.insert(hit.id)
            } else {
                unmatchedApp.append(txn)
            }
        }

        // Tier 2 — amount + date window
        let window = TimeInterval(dateWindowDays) * 86_400
        for txn in unmatchedApp.sorted(by: { $0.date < $1.date }) {
            let candidates = bank
                .filter { !consumedBank.contains($0.id)
                    && $0.amountPaise == txn.amountPaise
                    && $0.direction == txn.direction
                    && abs($0.date.timeIntervalSince(txn.date)) <= window }
                .sorted { lhs, rhs in
                    let dl = abs(lhs.date.timeIntervalSince(txn.date))
                    let dr = abs(rhs.date.timeIntervalSince(txn.date))
                    return dl == dr ? lhs.date < rhs.date : dl < dr
                }
            if let hit = candidates.first {
                matches.append(MatchPair(appID: txn.id, bankID: hit.id, tier: .amountDate))
                consumedBank.insert(hit.id)
            }
        }

        let matchedApp = Set(matches.map(\.appID))
        return ReconciliationResult(
            matches: matches,
            appUnmatched: app.filter { !matchedApp.contains($0.id) }.map(\.id),
            bankOnly: bank.filter { !consumedBank.contains($0.id) }.map(\.id)
        )
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/reconciliation.dart
/// Tiered reconciliation. Port of Reconciliation.swift.
library;

import 'domain.dart';

class ReconTxn {
  final String id;
  final DateTime date;
  final int amountPaise;
  final Direction direction;
  final String? reference;
  const ReconTxn({
    required this.id,
    required this.date,
    required this.amountPaise,
    required this.direction,
    required this.reference,
  });
}

enum MatchTier { reference, amountDate }

class MatchPair {
  final String appID;
  final String bankID;
  final MatchTier tier;
  const MatchPair(
      {required this.appID, required this.bankID, required this.tier});

  @override
  bool operator ==(Object other) =>
      other is MatchPair &&
      other.appID == appID &&
      other.bankID == bankID &&
      other.tier == tier;
  @override
  int get hashCode => Object.hash(appID, bankID, tier);
}

class ReconciliationResult {
  final List<MatchPair> matches;
  final List<String> appUnmatched;
  final List<String> bankOnly;
  const ReconciliationResult(
      {required this.matches,
      required this.appUnmatched,
      required this.bankOnly});
}

class Reconciler {
  /// Tier 1: exact reference match. Tier 2: same (amount, direction) within
  /// ±window days, closest date first; each bank txn consumed at most once.
  static ReconciliationResult reconcile({
    required List<ReconTxn> app,
    required List<ReconTxn> bank,
    int dateWindowDays = 2,
  }) {
    final matches = <MatchPair>[];
    final consumedBank = <String>{};
    final unmatchedApp = <ReconTxn>[];

    // Tier 1 — reference
    final bankByRef = <String, ReconTxn>{};
    for (final txn in bank) {
      final ref = txn.reference;
      if (ref != null && ref.isNotEmpty) {
        bankByRef.putIfAbsent(ref, () => txn);
      }
    }
    for (final txn in app) {
      final ref = txn.reference;
      final hit = (ref != null && ref.isNotEmpty) ? bankByRef[ref] : null;
      if (hit != null && !consumedBank.contains(hit.id)) {
        matches.add(
            MatchPair(appID: txn.id, bankID: hit.id, tier: MatchTier.reference));
        consumedBank.add(hit.id);
      } else {
        unmatchedApp.add(txn);
      }
    }

    // Tier 2 — amount + date window, closest first
    final window = Duration(days: dateWindowDays);
    unmatchedApp.sort((a, b) => a.date.compareTo(b.date));
    for (final txn in unmatchedApp) {
      final candidates = bank
          .where((b) =>
              !consumedBank.contains(b.id) &&
              b.amountPaise == txn.amountPaise &&
              b.direction == txn.direction &&
              b.date.difference(txn.date).abs() <= window)
          .toList()
        ..sort((l, r) {
          final dl = l.date.difference(txn.date).abs();
          final dr = r.date.difference(txn.date).abs();
          if (dl == dr) return l.date.compareTo(r.date);
          return dl.compareTo(dr);
        });
      if (candidates.isNotEmpty) {
        final hit = candidates.first;
        matches.add(
            MatchPair(appID: txn.id, bankID: hit.id, tier: MatchTier.amountDate));
        consumedBank.add(hit.id);
      }
    }

    final matchedApp = matches.map((m) => m.appID).toSet();
    return ReconciliationResult(
      matches: matches,
      appUnmatched:
          app.where((t) => !matchedApp.contains(t.id)).map((t) => t.id).toList(),
      bankOnly: bank
          .where((t) => !consumedBank.contains(t.id))
          .map((t) => t.id)
          .toList(),
    );
  }
}
